src/lib/optimizer/strategy/stored_table_column_alignment_rule.hpp

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+
+#include "abstract_rule.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+
+/*
+ The PQP sub plan memoization for StoredTableNodes is sensitive to the node's table name, set of pruned chunks and
+ set of pruned columns. Consequently, for multiple nodes with the same table name, same pruned chunks and same pruned
+ columns, only one GetTable operator is created and executed. In some queries, the ColumnPruningRule and
+ ChunkPruningRule provide an LQP with multiple StoredTableNodes where the table names and sets of pruned chunks are
+ equal but the sets of pruned columns are different. This leads to the creation and execution of different GetTable
+ operators instead of only one operator.
+
+ This rule identifies StoredTableNodes that differ only in the pruned columns. It then intersects the lists of pruned
+ columns. While this means that some columns are left unpruned, it makes the job easier for the memoization in the
+ LQPTranslator. In our experiments, this has led to significant performance improvements and negligible reductions.
+
+ Example: LQP sub plan before executing the StoredTableColumnAlignmentRule
+
+                      +------------------------+
+                      | Join (1)               |
+                      | Mode: Inner            |
+                      | predicate: Li_a = Pa_a |
+                      +------------------------+
+                         /                \
+       +------------------------+     +------------------------+
+       | Join (2)               |     | Join (3)               |
+       | Mode: Semi             |     | Mode: Semi             |
+       | predicate: Li_a = Pa_a |     | predicate: Li_a = Pa_a |
+       +------------------------+     +------------------------+
+                   /          \           /           \
+ +---------------------+  +-----------------------+  +-----------------------+
+ | StoredTable (1)     |  | StoredTable (2)       |  | StoredTable (3)       |
+ | table name:     Li  |  | table name:     Pa    |  | table name:     Li    |
+ | pruned chunks:  { } |  | pruned chunks:  {1,2} |  | pruned chunks:  { }   |
+ | pruned columns: {3} |  | pruned columns: { }   |  | pruned columns: {3,4} |
+ +---------------------+  +-----------------------+  +-----------------------+
```

Comment:
Are you sure the rule applies here? I was tempted to make this change:

```
       +------------------------+     +------------------------+
       | Join (2)               |     | Join (3)               |
       | Mode: Semi             |     | Mode: Semi             |
       | predicate: Li_a = Pa_a |     | predicate: Li_a = Pa_a |
       +------------------------+     +------------------------+
             left /          \   right   /           \ left
```

but that would mean that table Li is on the left side of the semi join in both cases. As a result, Pa would not be available for the final join.

If, however, the left and right inputs for join (2) and (3) were as they are in the image, it would mean that the joins would not compare equal even after the rule is applied.

Reply:
The example was inspired by TPC-H Q17 and the lineitem table was at the left side for both join nodes. So yes, your descriptions of the sides are correct but it's also true that Join (1) is not correct then. How about changing Join (2) and (3) to Inner and Join (1) to Semi?